### PR TITLE
Bug 2059700: [4.9z] After reboot egress node, lr-policy-list was not correct, some duplicate records or missed internal IPs

### PR DESCRIPTION
**Pull request by flavio-fernandes**
Use pod IPs from annotation in sync

We cannot rely on pod.Status.PodIPs for obtaining the IPs, since that is unset in between the restart and the moment it retrieves that information. Use annotation instead.

In cases where OVN database for logical router policies or NATs used by EgressIPs have stale nexthops or wrong external_ips, the sync function should remove them, so the proper row/column gets set.

Signed-off-by: Flavio Fernandes [flaviof@redhat.com](mailto:flaviof@redhat.com)

### CODE REVIEW

**Comment by flavio-fernandes** [comment]
/hold
need backport changes from https://github.com/openshift/ovn-kubernetes/pull/994

**Comment by flavio-fernandes** [comment]
/remove-hold
@kyrtapz @trozet PTAL. The backport of non-libovsdb may require some additional time. :)

**Comment by flavio-fernandes** [comment]
Note this is a backport from 4.10's PR:  https://github.com/openshift/ovn-kubernetes/pull/994 ... for the most part.

File: go-controller/pkg/ovn/egressip.go [hunk 1]
```diff
@@ -357,19 +391,35 @@ func (oc *Controller) syncStaleNATRules(egressIPToPodIPCache map[string]sets.Str
...
-// generatePodIPCacheForEgressIP builds a cache of egressIP name -> podIPs for fast
+// generateCacheForEgressIP builds a cache of egressIP name -> podIPs for fast
 // access when syncing egress IPs. The Egress IP setup will return a lot of
 // atomic items with the same general information repeated across most (egressIP
 // name, logical IP defined for that name), hence use a cache to avoid round
 // trips to the API server per item.
-func (oc *Controller) generatePodIPCacheForEgressIP(eIPs []interface{}) (map[string]sets.String, error) {
-	egressIPToPodIPCache := make(map[string]sets.String)
+func (oc *Controller) generateCacheForEgressIP(eIPs []interface{}) (map[string]egressIPCacheEntry, error) {
+	// egressIPToPodIPCache := make(map[string]sets.String)
```
Comment (trozet) [inline go-controller/pkg/ovn/egressip.go:400]: remove

File: go-controller/pkg/ovn/egressip.go [hunk 2]
```diff
@@ -845,7 +895,7 @@ func (e *egressIPController) addPodEgressIP(eIP *egressipv1.EgressIP, pod *kapi.
 	if e.needsRetry(pod) {
 		e.podRetry.Delete(getPodKey(pod))
 	}
-	if err := e.handleEgressReroutePolicy(podIPs, eIP.Status.Items, eIP.Name, e.createEgressReroutePolicy); err != nil {
+	if err := e.handleEgressReroutePolicy(podIPs, eIP.Status.Items, eIP.Name, e.createOrUpdateEgressReroutePolicy); err != nil {
```
Comment (trozet) [inline go-controller/pkg/ovn/egressip.go:898]: the changes from here to the end of this file do not correspond to the cherry-picked commit. I think you need to split that into a separate commit and list the cherry-picked commit it came from or if its not something cherry picked and new code downstream only for 4.9z then should be a comiit that starts with CARRY: <fix idempotently updating egress IP route policies" or something...

File: go-controller/pkg/ovn/egressip.go [hunk 3]
```diff
@@ -991,50 +1042,102 @@ func (e *egressIPController) createEgressReroutePolicy(filterOption, egressIPNam
...
 			return fmt.Errorf("unable to create logical router policy, stderr: %s, err: %v", stderr, err)
 		}
+		return nil
+	}
+
+	if logicalRoutePoliciesLen > 1 {
+		klog.Warningf("Found %d OVN rows for egressIP %s %s. ovnkube-master restart may fix it.", logicalRoutePoliciesLen, egressIPName, filterOption)
+	}
+
+	// At this point we know there is a policy for egressIPName with the wanted filterOption. Make sure its nextops are proper.
```
Comment (trozet) [inline go-controller/pkg/ovn/egressip.go:1052]: nexthops

File: go-controller/pkg/ovn/egressip.go [hunk 4]
```diff
@@ -991,50 +1042,102 @@ func (e *egressIPController) createEgressReroutePolicy(filterOption, egressIPNam
...
+	expectedNexthops := sets.NewString()
+	for _, gatewayRouterIP := range gatewayRouterIPs {
+		expectedNexthops.Insert(string(gatewayRouterIP.String()))
+	}
+	if !expectedNexthops.Equal(logicalRoutePolicies[0].nexthops) {
+		klog.Infof("createOrUpdateEgressReroutePolicy will update nexthops of %s from: %v to: %v", egressIPName, logicalRoutePolicies[0].nexthops.List(), expectedNexthops.List())
+		err = e.updateEgressReroutePolicyNexthops(logicalRoutePolicies[0].policyID, expectedNexthops.List())
```
Comment (trozet) [inline go-controller/pkg/ovn/egressip.go:1059]: nit: expectedNexthops.List() is slightly slower than UnsortedList due to the sorting it has to do

**Review by trozet — COMMENTED** [review]
some minor nits but overall looks good

File: go-controller/pkg/ovn/egressip.go (hunk 2 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go:897]: done.

File: go-controller/pkg/ovn/egressip.go [hunk 5]
```diff
@@ -259,6 +265,10 @@ func (oc *Controller) syncStaleEgressReroutePolicy(egressIPToPodIPCache map[stri
 		klog.Errorf("Unable to sync egress IPs, unable to find logical router policies, stderr: %s, err: %v", stderr, err)
 		return
 	}
+
+	// Set of visited podIps per egressIPName, so duplicates can be detected and removed
+	podIPsVisited := make(map[string]sets.String)
```
[the remark this answers is not in the document]
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go]: done.

File: go-controller/pkg/ovn/egressip.go (hunk 1 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go]: done.

File: go-controller/pkg/ovn/egressip.go (hunk 3 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip.go]: done.

File: go-controller/pkg/ovn/egressip.go [hunk 6]
```diff
@@ -1040,64 +1041,103 @@ func (e *egressIPController) createEgressReroutePolicy(filterOption, egressIPNam
...
 		}
 	}
+	_, stderr, err := txn.Commit()
+	if err != nil {
+		return fmt.Errorf("unable to remove logical router policies, stderr: %s, err: %v", stderr, err)
+	}
 	return nil
 }
 
-func findReroutePolicyIDs(filterOption, egressIPName string, gatewayRouterIPs []net.IP) ([]string, error) {
```
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/egressip.go:1009]: this is a bit confusing... updateEgressReroutePolicyNexthops did not change with this commit. diff tool is a little lost... :)
